bazel_repositories

```diff
@@ -0,0 +1,494 @@
+    whether the dependency is excluded and should be removed later.
+    """
+    filename = pathlib.Path(urllib_parse.urlparse(url).path).parts[-1]
+    # Sometimes the same dependency might be listed multiple times and have
+    # the same filename. To avoid conflicts, append a random UUID.
+    filename = f"{filename}-{uuid.uuid4()}"
+    try:
+        download(url, filename)
+    except (urllib_error.HTTPError, urllib_error.URLError):
+        log.warning(f"could not download {url}")
+        return
+
+    log.info(f"Checking sha256 sum of {url}")
+    sha256sum = hashlib.sha256()
+    with open(filename, "rb") as f:
+        while True:
+            buf = f.read(BUF_SIZE)
+            if not buf:
+                break
+            sha256sum.update(buf)
+    sha256sum_digest = sha256sum.hexdigest()
+
+    # Sometimes we might have the same tarball listed in dependencies twice,
+    # sometimes even under different names. In that case, that tarball is going
+    # to be stored in the same `BAZEL_CACHE/<sha256sum>/file` location multiple
+    # times. This lock ensures no conflicts in moving files to `BAZEL_CACHE`.
+    lock.acquire()
+
+    log.info(f"Storing {url}")
+    os.makedirs(os.path.join(CACHEDIR, sha256sum_digest), exist_ok=True)
+    cache_filename = os.path.join(CACHEDIR, sha256sum_digest, "file")
+    shutil.move(filename, cache_filename)
+
+    lock.release()
+
+    excluded = False
+    for e in exclude:
+        if e in url:
+            log.info(f"excluding dependency {url}")
+            excluded = True
+
+    return url, cache_filename, excluded
+
+
+def process_urls(urls: typing.List[str], exclude: typing.List[str]) \
+        -> typing.List[str]:
+    """Download and store archives from the given list of URLs. Return the list
+    of tuples with three values: URL, path of th archive in thee cache and
+    boolean value whether the dependency is excluded and should be removed
+    later.
+    """
+    # Remove duplicates
+    urls = list(set(urls))
+
+    pool = multiprocessing.Pool()
+    manager = multiprocessing.Manager()
+    lock = manager.Lock()
+    func = functools.partial(process_url, lock, exclude)
+    processed_urls = pool.imap(func, urls)
+    processed_urls = [x for x in processed_urls if x is not None]
+
+    return processed_urls
+
+
+def compress_cache():
+    """Compress the Bazel cache dir into a tarball."""
+    with tarfile.open("vendor.tar.gz", "w:gz") as tar:
+        tar.add(CACHEDIR)
+    shutil.rmtree(CACHEDIR)
+
+
+def clean_spec():
+    """Clean the spec file from autogenerated content. This step is needed
+    before running quilt to ensure that it will only use the main source
+    tarball and not the vendor tarball produced by this service.
+    """
+    content = ""
+    ignore = False
+    log.info("cleaning spec file")
+    with codecs.open(find_spec(), "r+", "utf-8") as f:
+        for line in f:
+            if line.startswith(AUTOGEN_FOOTER):
```

Comment:
Should this be checking for ignore as well? I guess the odds of a spec file having the footer but no header are pretty unlikely.

Reply:
It doesn't harm to add a check for `ignore` as well. I will do that.

Reply:
I will change that to:

```diff
diff --git a/bazel_repositories b/bazel_repositories
index 2e2ac2b..2bae822 100755
--- a/bazel_repositories
+++ b/bazel_repositories
@@ -410,10 +410,9 @@ def update_spec(urls: typing.List[str]):
     log.info("updating spec file")
     with codecs.open(find_spec(), "r+", "utf-8") as f:
         for line in f:
-            if line.startswith(AUTOGEN_FOOTER):
-                ignore = False
-                continue
             if ignore:
+                if line.startswith(AUTOGEN_FOOTER):
+                    ignore = False
                 continue
             if line.startswith(AUTOGEN_HEADER):
                 ignore = True
```

Reply:
done